src/tslintMulti.ts

```diff
@@ -0,0 +1,155 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ts from "typescript";
+
+import {
+    DEFAULT_CONFIG,
+    findConfiguration,
+    findConfigurationPath,
+    getRelativePath,
+    getRulesDirectories,
+    loadConfigurationFromPath,
+} from "./configuration";
+import { EnableDisableRulesWalker } from "./enableDisableRules";
+import { findFormatter } from "./formatterLoader";
+import { IFormatter } from "./language/formatter/formatter";
+import { RuleFailure } from "./language/rule/rule";
+import { TypedRule } from "./language/rule/typedRule";
+import { getSourceFile } from "./language/utils";
+import { IMultiLinterOptions, LintResult } from "./lint";
+import { loadRules } from "./ruleLoader";
+import { arrayify } from "./utils";
+
+/**
+ * Linter that can lint multiple files in consecutive runs.
+ */
+class MultiLinter {
+
+    public static VERSION = "3.14.0";
+
+    public static findConfiguration = findConfiguration;
+    public static findConfigurationPath = findConfigurationPath;
+    public static getRulesDirectories = getRulesDirectories;
+    public static loadConfigurationFromPath = loadConfigurationFromPath;
+
+    private failures: RuleFailure[] = [];
+
+    /**
+     * Creates a TypeScript program object from a tsconfig.json file path and optional project directory.
+     */
+    public static createProgram(configFile: string, projectDirectory?: string): ts.Program {
+        if (projectDirectory === undefined) {
+            const lastSeparator = configFile.lastIndexOf("/");
+            if (lastSeparator < 0) {
+                projectDirectory = ".";
+            } else {
+                projectDirectory = configFile.substring(0, lastSeparator + 1);
+            }
+        }
+
+        const {config} = ts.readConfigFile(configFile, ts.sys.readFile);
+        const parsed = ts.parseJsonConfigFileContent(config, {readDirectory: ts.sys.readDirectory}, projectDirectory);
+        const host = ts.createCompilerHost(parsed.options, true);
+        const program = ts.createProgram(parsed.fileNames, parsed.options, host);
+
+        return program;
+    }
+
+    /**
+     * Returns a list of source file names from a TypeScript program. This includes all referenced
+     * files and excludes declaration (".d.ts") files.
+     */
+    public static getFileNames(program: ts.Program): string[] {
+        return program.getSourceFiles().map(s => s.fileName).filter(l => l.substr(-5) !== ".d.ts");
+    }
+
+    constructor(private options: IMultiLinterOptions, private program?: ts.Program) {
+        // Empty
+    }
+
+    public lint(fileName: string, source: string, configuration: any = DEFAULT_CONFIG): void {
```

Comment:
> Would you mind if we do this in another PR? This change would probably require a couple of other changes (including to the test setup), so I tend to think it would be more safe to separate this.

Yep, makes sense to me, that seems fine.

> Actually, I would prefer if we could omit the `#lint` method's `configuration` parameter at all. It would therefore be cleaner to attach the whole configuration including options (by merging these) to the instance of the linter.

I'm fine with this as well - it seems like a reasonable breaking change to include as part of 4.0. We have to be careful and consider all the ways TSLint gets used - as parts of IDE addons, as part of build tools, from the CLI, etc. But I think your ideas are solid - an IDE plugin could create a new instance of the linter for linting each file one by one as it's open; a build tool can lint all files together; etc.